server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -4621,10 +4639,81 @@ public void syncSegmentsFromRemoteSegmentStore(boolean overrideLocal, boolean re
+     * @param commitGeneration commit generation at the time of commit operation for which we are syncing segments
+     * @throws IOException if exception occurs while reading segments from remote store
+     */
+    public void syncSegmentsFromGivenRemoteSegmentStore(
+        boolean overrideLocal,
+        RemoteSegmentStoreDirectory sourceRemoteSegmentDirectory,
+        long primaryTerm,
+        long commitGeneration
+    ) throws IOException {
+        logger.info("Downloading segments from given remote segment store");
+        RemoteSegmentStoreDirectory remoteSegmentStoreDirectory = null;
+        if (remoteStore != null) {
+            remoteSegmentStoreDirectory = getRemoteSegmentDirectoryForShard();
+            remoteSegmentStoreDirectory.init();
+            remoteStore.incRef();
+        }
+        sourceRemoteSegmentDirectory.init();
+        Map<String, RemoteSegmentStoreDirectory.UploadedSegmentMetadata> uploadedSegments = sourceRemoteSegmentDirectory
+            .getSegmentsUploadedToRemoteStore(primaryTerm, commitGeneration);
+        final Directory storeDirectory = store.directory();
+        store.incRef();
+
+        try {
+            copySegmentFiles(storeDirectory, sourceRemoteSegmentDirectory, remoteSegmentStoreDirectory, uploadedSegments, overrideLocal);
+        } catch (IOException e) {
+            throw new IndexShardRecoveryException(shardId, "Exception while copying segment files from remote segment store", e);
+        } finally {
+            store.decRef();
+            if (remoteStore != null) {
+                remoteStore.incRef();
+            }
+        }
+    }
+
+    private void copySegmentFiles(
+        Directory storeDirectory,
+        RemoteSegmentStoreDirectory sourceRemoteDirectory,
+        RemoteSegmentStoreDirectory targetRemoteDirectory,
+        Map<String, RemoteSegmentStoreDirectory.UploadedSegmentMetadata> uploadedSegments,
+        boolean overrideLocal
+    ) throws IOException {
+        List<String> downloadedSegments = new ArrayList<>();
+        List<String> skippedSegments = new ArrayList<>();
+        try {
+            Set<String> localSegmentFiles = Sets.newHashSet(storeDirectory.listAll());
+            for (String file : uploadedSegments.keySet()) {
+                long checksum = Long.parseLong(uploadedSegments.get(file).getChecksum());
+
+                if (overrideLocal || localDirectoryContains(storeDirectory, file, checksum) == false) {
+                    if (localSegmentFiles.contains(file)) {
+                        storeDirectory.deleteFile(file);
+                    }
+                    storeDirectory.copyFrom(sourceRemoteDirectory, file, file, IOContext.DEFAULT);
+                    storeDirectory.sync(Collections.singleton(file));
+                    if (targetRemoteDirectory != null) {
+                        targetRemoteDirectory.copyFrom(storeDirectory, file, file, IOContext.DEFAULT);
```

Comment:
We need to copy to `targetRemoteDirectory` irrespective of whether we are skipping the download to local fs.